### PR TITLE
Nouns 🤝 i18n

### PR DESCRIPTION
**Pull request by bcjgit**
# TLDR
This PR does the following:

1. Adds i18n support via Lingui.js  for nearly all strings in nouns.wtf (all but a handful of strings that were deemed low priority + inside of react components that made them hard to wrap in the `<Trans>` macro were translated)
2. Add UI (mobile and desktop) to allow users to switch between Languages
3. Add logic on the backend to deal with said switching (based heavily off of the implementation in `Uniswap/interface`
4. Add logic to detect user's system / browser locale (again, based off Uniswap) 
5. Adds basic support for Japanese (strings and _mostly_ dummy right now but final translations should be coming in next ~48 hours - have a call with translator early morning US time 4/27)

## Testing Done
1. Played around with language switching on desktop and mobile (both my actual phone and chrome sim) and confirmed it worked as expected
2. Confirmed that locale was auto-detected via hardcoding my system language (safari and chrome) 
3. Visually confirmed via pseudolocalization that (nearly) all strings have been wrapped in translation logic

### CODE REVIEW

File: packages/nouns-webapp/src/components/NavLocaleSwitcher/index.tsx [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+import { faGlobe } from '@fortawesome/free-solid-svg-icons';
+import { Dropdown } from 'react-bootstrap';
+import clsx from 'clsx';
+import { useHistory } from 'react-router-dom';
+import { isMobileScreen } from '../../utils/isMobile';
+import { usePickByState } from '../../utils/colorResponsiveUIUtils';
+import LanguageSelectionModal from '../LanguageSelectionModal';
+import { setLocale } from '../../i18n/setLocale';
+import { Trans } from '@lingui/macro';
+import navWalletClasses from '../NavWallet/NavWallet.module.css';
```
Comment (solimander) [inline packages/nouns-webapp/src/components/NavLocaleSwitcher/index.tsx:16]: Should rename this if used by both the nav wallet and language picker

**Comment by solimander** [comment]
Language button + dropdown isn't aware of screen width changes. If you load at a small screen size and then widen, it still behaves as if it's on mobile (and vice versa).



e.g.

desktop

mobile

File: packages/nouns-webapp/src/i18n/setLocale.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+ * Sets locale in local storage
+ *
+ * Note: this value will persist across sessions
+ *
+ * @param locale Locale we wish to use for this user
+ */
+export const setLocale = (locale: string) => {
+  if (localStorage.getItem('lang') !== locale) {
+    localStorage.setItem('lang', locale);
+    window.location.reload();
```
Comment (solimander) [inline packages/nouns-webapp/src/i18n/setLocale.ts:13]: What's the purpose of the reload? Shouldn't be required to do a full page refresh on locale change.

File: packages/nouns-webapp/package.json [hunk 3]
```diff
@@ -66,7 +71,11 @@
     "start": "react-scripts start",
     "build": "react-scripts build",
     "test": "react-scripts test",
-    "eject": "react-scripts eject"
+    "eject": "react-scripts eject",
+    "i18n:extract": "lingui extract",
+    "i18n:compile": "lingui compile",
+    "i18n:pseudo": "lingui extract --locale pseudo && lingui compile"
+
```
Comment (solimander) [inline packages/nouns-webapp/package.json:78]: Nit: Remove unused line

File: packages/nouns-webapp/src/components/Bid/index.tsx [hunk 4]
```diff
@@ -162,27 +167,35 @@ const Bid: React.FC<{
...
       case 'Fail':
         setModal({
-          title: 'Transaction Failed',
-          message: placeBidState.errorMessage ? placeBidState.errorMessage : 'Please try again.',
+          title: <Trans>Transaction Failed</Trans>,
+          message: placeBidState.errorMessage ? (
+            placeBidState.errorMessage
+          ) : (
+            <Trans>Please try again.</Trans>
+          ),
```
Comment (solimander) [inline packages/nouns-webapp/src/components/Bid/index.tsx:183]: I know this was already in place, but can make these more readable by using optional chaining and or operator.

e.g.

```ts
placeBidState?.errorMessage || <Trans>Please try again.</Trans>,
```

File: packages/nouns-webapp/src/components/Bid/index.tsx [hunk 5]
```diff
@@ -276,7 +298,7 @@ const Bid: React.FC<{
           <>
             <Col lg={12} className={classes.voteForNextNounBtnWrapper}>
               <Button className={classes.bidBtnAuctionEnded} onClick={fomoNounsBtnOnClickHandler}>
-                Vote for the next Noun ⌐◧-◧
+                <Trans>Vote for the next Noun ⌐◧-◧</Trans>
```
Comment (solimander) [inline packages/nouns-webapp/src/components/Bid/index.tsx:301]: Exclude `⌐◧-◧` from translations

File: packages/nouns-webapp/src/components/NavLocaleSwitcher/NavLocalSwitcher.module.css [hunk 6]
```diff
@@ -0,0 +1,70 @@
+.desktopLanguageButton {
```
Comment (solimander) [inline packages/nouns-webapp/src/components/NavLocaleSwitcher/NavLocalSwitcher.module.css:1]: I don't believe this file is used. If wallet styles are sufficient then generalize those for use here as well.

File: packages/nouns-webapp/src/locales/en-US.js [hunk 7]
```diff
@@ -0,0 +1,317 @@
+/*eslint-disable*/ module.exports = {
```
Comment (solimander) [inline packages/nouns-webapp/src/locales/en-US.js:1]: The build catalogs typically should not be committed. Add to the `.gitignore` and compile in the `prepare` hook instead.

File: packages/nouns-webapp/src/i18n/setLocale.ts (hunk 2 above)
Comment (bcjgit) [reply packages/nouns-webapp/src/i18n/setLocale.ts:13]: Could be missing something, but I found that the date localization didn't work until after a reload. Less importantly, this was also to fix issues with a few locale specific UI changes I made (spacing etc.) that look at `useActiveLocale` to trigger.

File: packages/nouns-webapp/src/components/NavLocaleSwitcher/NavLocalSwitcher.module.css (hunk 6 above)
Comment (bcjgit) [reply packages/nouns-webapp/src/components/NavLocaleSwitcher/NavLocalSwitcher.module.css:1]: It's used here:
https://sourcegraph.com/github.com/nounsDAO/nouns-monorepo@brianj_add_i18n_support_plus_ja_support/-/blob/packages/nouns-webapp/src/components/NavLocaleSwitcher/index.tsx?L128

I agree though that wallet dropdown styles are used in both places so will rename those to be more general

Comment (bcjgit) [reply packages/nouns-webapp/src/components/NavLocaleSwitcher/NavLocalSwitcher.module.css:1]: done

**Comment by bcjgit** [comment]
@solimander update 5/1:

1. Refactored `NavWallet.module.css` into 2 classfiles (`NavWallet.module.css` and `NavBarDropdown.module.css`). The cross-cutting classes now being shared in the later per your comment. 
2.  Added build category to `.gitignore` and updated build to generate them on the fly accordingly
3. Excluded noggles from translation 
4. Updated error copy logic in `Bid.tsx` to update the `||` operator per your comment
5. Updated responsiveness logic 
6. Worked with translator to make a few minor changes to copy based on seeing them in the full context of the page
7. Made a handful of UI tweaks for things don't look strange in `ja_JP`

File: packages/nouns-webapp/src/i18n/setLocale.ts (hunk 2 above)
Comment (solimander) [reply packages/nouns-webapp/src/i18n/setLocale.ts:13]: I'll take a look at the date localization issue, but it'd be a much nicer UX if we didn't need to do a full-page refresh.

Comment (solimander) [reply packages/nouns-webapp/src/i18n/setLocale.ts:13]: Enabling `forceRenderOnLocaleChange` should solve the same issue and provide a cleaner UX. It will re-render the UI without requiring API data to be re-fetched.

Comment (bcjgit) [reply packages/nouns-webapp/src/i18n/setLocale.ts:13]: I had that enabled (I think) but still required a full reload to get the dates to change ... will also look into this. Fully ack that I could just be being stupid on this...

File: packages/nouns-webapp/package.json [hunk 8]
```diff
@@ -64,9 +69,12 @@
   },
   "scripts": {
     "start": "react-scripts start",
-    "build": "react-scripts build",
+    "build": "yarn i18n:compile && react-scripts build",
     "test": "react-scripts test",
-    "eject": "react-scripts eject"
+    "eject": "react-scripts eject",
+    "i18n:extract": "lingui extract",
+    "i18n:compile": "lingui compile",
```
Comment (solimander) [inline packages/nouns-webapp/package.json:76]: Add `"prepare": "yarn i18n:compile",` so localization works out of the box after installing dependencies for local development.

**Comment by solimander** [comment]
nit: run `yarn format`

**Review by solimander — APPROVED** [review]
🇯🇵

**Comment by KSST88** [comment]
> # TLDR
> This PR does the following:
> 
> 1. Adds i18n support via Lingui.js  for nearly all strings in nouns.wtf (all but a handful of strings that were deemed low priority + inside of react components that made them hard to wrap in the `<Trans>` macro were translated)
> 2. Add UI (mobile and desktop) to allow users to switch between Languages
> 3. Add logic on the backend to deal with said switching (based heavily off of the implementation in `Uniswap/interface`
> 4. Add logic to detect user's system / browser locale (again, based off Uniswap)
> 5. Adds basic support for Japanese (strings and _mostly_ dummy right now but final translations should be coming in next ~48 hours - have a call with translator early morning US time 4/27)
> 
> ## Testing Done
> 1. Played around with language switching on desktop and mobile (both my actual phone and chrome sim) and confirmed it worked as expected
> 2. Confirmed that locale was auto-detected via hardcoding my system language (safari and chrome)
> 3. Visually confirmed via pseudolocalization that (nearly) all strings have been wrapped in translation logic

**Comment by KSST88** [comment]
> # TLDR
> This PR does the following:
> 
> 1. Adds i18n support via Lingui.js  for nearly all strings in nouns.wtf (all but a handful of strings that were deemed low priority + inside of react components that made them hard to wrap in the `<Trans>` macro were translated)
> 2. Add UI (mobile and desktop) to allow users to switch between Languages
> 3. Add logic on the backend to deal with said switching (based heavily off of the implementation in `Uniswap/interface`
> 4. Add logic to detect user's system / browser locale (again, based off Uniswap)
> 5. Adds basic support for Japanese (strings and _mostly_ dummy right now but final translations should be coming in next ~48 hours - have a call with translator early morning US time 4/27)
> 
> ## Testing Done
> 1. Played around with language switching on desktop and mobile (both my actual phone and chrome sim) and confirmed it worked as expected
> 2. Confirmed that locale was auto-detected via hardcoding my system language (safari and chrome)
> 3. Visually confirmed via pseudolocalization that (nearly) all strings have been wrapped in translation logic
